### ISSUE TITLE
Inverter Protocol

### ISSUE THREAD

**Issue by kmsj13** [TradeFace/modbusmqtt#1]
Does Inverter uses RS232 or RS485?

**Comment by TradeFace** [comment]
My inverters communicate over zigbee to the ecd. The ecd has a rs485 port which connects to the WiFi RTU. I intercept the messages pushed by the RTU. These messages are indeed in RS485 format

**Comment by kmsj13** [comment]
What is the model of your existing RTU? i' am looking for possibility that your code might be work on SUN GTIL1000G2 (https://sunshine.en.made-in-china.com/product/uKSQCLnUnXcs/China-1kw-Pure-Sine-Wave-Inverter-Grid-Tie-Inverter-Inverter-SUN-1000G2-LCD-.html)

**Comment by TradeFace** [comment]
I think it is HF-A21

System Info
--
Software Version | 2.0.7.0
Hardware Version | 2.0.1.0
MAC | F0:FE:6B:88:85:00

**Comment by kmsj13** [comment]
Alright, do you have some documentation regarding registers of DEYE Inverter?

**Comment by TradeFace** [comment]
Nope, I reverse engineered it. All (limited) knowledge is in the files posted here

**Comment by kmsj13** [comment]
I see, i read some of your codes and found that there is some commands sending it to your RTU, does your RTU waits command from modbus server and relay it to UART of ECD then the reply message from UART send back it?

**Comment by TradeFace** [comment]
my solar setup: panels -> microinverters --zigbee--> ECD --RS485--> Wifi RTU -- wifi intercept --> modbusmqtt --json--> mqtt

modbusmqtt acts as a modbus server. So the RTU automatically connects to the modbus server. 
I had to set www.shinemonitor.com to the local IP of modbusmqtt in my router. 

Does that answer your question?

**Comment by kmsj13** [comment]
does modbusmqtt sends request command to your WIFI RTU to gather data or is it the WIFI RTU sends data intervally to your modbusmqtt without any request coming from it?

**Comment by TradeFace** [comment]
WIFI RTU sends data intervally to modbusmqtt without any request, indeed

**Comment by TradeFace** [comment]
The Wifi RTU connects by it self to modbusmqtt. If I remember correctly, the RTU expects a acknowledgement with a transaction number, so it knows its connected. And will reply to that with a status update. (I am by no means an expert on the RS485 tcp protocol... this was my first contact with it, so I might be way off here)

In the config file there is an option to switch debugging on. I think it will show all the communication between modbusmqtt and Wifi RTU.

**Comment by vosjevos** [comment]
I am watching this interesting topic as I am working on a solution to grab the solar panel data **directly from the RS485 interface** (so bypassing the WiFi-Plug) and feed it into my Raspberry Pi running Domoticz.  

My PV installation consists of 3 microinverters SUN1200G that communicate via PLC to the Deye SUN ECD box with WiFI plug. See diagrams. Does anybody has more documentation on the messages and coding?

Currently I fetch the data on a 5 minute basis from shine monitor.com.

![schema](https://user-images.githubusercontent.com/65259273/82117257-3d126080-976f-11ea-9435-090759c8a3fb.png)

**Comment by TradeFace** [comment]
@vosjevos That looks similar to my setup, though in my case the ecd gets its data via zigbee from the inverters. I have the ECD connected to the Wifi RTU via RS 485/modbus. What I know about the encoding/protocol is put into the code... (ben gestopt toen het werkte ;) )

**Comment by vosjevos** [comment]
How often do you fetch the data? Every 5 minutes?

**Comment by TradeFace** [comment]
I have it set at every 60s. (could be more, but have no use for finer grain of data at the moment)

**Comment by kmsj13** [comment]
Do you know the pinout for RS485 port of ECD?

**Comment by melaz** [comment]
Can you please help me understand how you made mapping for device? I'm trying to setup my, but cant figure out how you did mapping.

**Comment by TradeFace** [comment]
Mapping was done in https://github.com/TradeFace/modbusmqtt/blob/master/devices/shine_17.yaml

You'll have to look at the raw output of the device to see which data value is set at which offset and for how many "characters". Compare it to known values to find which value is represented at a register. Trail and error
